File: backend/app/models/defensia.py
"""DefensIA — Pydantic models y enums.

Spec: plans/2026-04-13-defensia-design.md §5.2, §7.4
"""
from __future__ import annotations
from datetime import datetime
from enum import Enum
from typing import Any, Optional
from pydantic import BaseModel, Field, ConfigDict


class Tributo(str, Enum):
    IRPF = "IRPF"
    IVA = "IVA"
    ISD = "ISD"
    ITP = "ITP"
    PLUSVALIA = "PLUSVALIA"


class Fase(str, Enum):
    COMPROBACION_REQUERIMIENTO = "COMPROBACION_REQUERIMIENTO"
    COMPROBACION_PROPUESTA = "COMPROBACION_PROPUESTA"
    COMPROBACION_POST_ALEGACIONES = "COMPROBACION_POST_ALEGACIONES"
    LIQUIDACION_FIRME_PLAZO_RECURSO = "LIQUIDACION_FIRME_PLAZO_RECURSO"
    SANCIONADOR_INICIADO = "SANCIONADOR_INICIADO"
    SANCIONADOR_PROPUESTA = "SANCIONADOR_PROPUESTA"
    SANCIONADOR_IMPUESTA = "SANCIONADOR_IMPUESTA"
    REPOSICION_INTERPUESTA = "REPOSICION_INTERPUESTA"
    TEAR_INTERPUESTA = "TEAR_INTERPUESTA"
    TEAR_AMPLIACION_POSIBLE = "TEAR_AMPLIACION_POSIBLE"
    FUERA_DE_ALCANCE = "FUERA_DE_ALCANCE"
    INDETERMINADA = "INDETERMINADA"


class TipoDocumento(str, Enum):
    REQUERIMIENTO = "REQUERIMIENTO"
    PROPUESTA_LIQUIDACION = "PROPUESTA_LIQUIDACION"
    LIQUIDACION_PROVISIONAL = "LIQUIDACION_PROVISIONAL"
    ACUERDO_INICIO_SANCIONADOR = "ACUERDO_INICIO_SANCIONADOR"
    PROPUESTA_SANCION = "PROPUESTA_SANCION"
    ACUERDO_IMPOSICION_SANCION = "ACUERDO_IMPOSICION_SANCION"
    ESCRITO_ALEGACIONES_USUARIO = "ESCRITO_ALEGACIONES_USUARIO"
    ESCRITO_REPOSICION_USUARIO = "ESCRITO_REPOSICION_USUARIO"
    ESCRITO_RECLAMACION_TEAR_USUARIO = "ESCRITO_RECLAMACION_TEAR_USUARIO"
    ACTA_INSPECCION = "ACTA_INSPECCION"
    PROVIDENCIA_APREMIO = "PROVIDENCIA_APREMIO"
    RESOLUCION_TEAR = "RESOLUCION_TEAR"
    RESOLUCION_TEAC = "RESOLUCION_TEAC"
    SENTENCIA_JUDICIAL = "SENTENCIA_JUDICIAL"
    JUSTIFICANTE_PAGO = "JUSTIFICANTE_PAGO"
    FACTURA = "FACTURA"
    ESCRITURA = "ESCRITURA"
    LIBRO_REGISTRO = "LIBRO_REGISTRO"
    OTROS = "OTROS"


class EstadoExpediente(str, Enum):
    BORRADOR = "borrador"
    EN_ANALISIS = "en_analisis"
    DICTAMEN_LISTO = "dictamen_listo"
    ARCHIVADO = "archivado"


class DocumentoEstructurado(BaseModel):
    model_config = ConfigDict(use_enum_values=False)

    id: str
    nombre_original: str
    tipo_documento: TipoDocumento
    fecha_acto: Optional[datetime] = None
    datos: dict[str, Any] = Field(default_factory=dict)
    clasificacion_confianza: float = 1.0


class Brief(BaseModel):
    id: Optional[str] = None
    texto: str
    chat_history: list[dict[str, str]] = Field(default_factory=list)


class ExpedienteEstructurado(BaseModel):
    model_config = ConfigDict(use_enum_values=False)

    id: str
    tributo: Tributo
    ccaa: str
    documentos: list[DocumentoEstructurado] = Field(default_factory=list)
    fase_detectada: Fase = Fase.INDETERMINADA
    fase_confianza: float = 0.0

    def timeline_ordenado(self) -> list[DocumentoEstructurado]:
        """Devuelve documentos ordenados ASC por fecha_acto.

        Los documentos sin fecha_acto se colocan al final preservando el orden
        relativo de entrada (sort estable). Normaliza naive→UTC para evitar
        TypeError al mezclar fechas naive/aware (Gemini/parseo PDF puede
        devolver ambos). NO muta los documentos originales.
        """
        from datetime import timezone as _tz

        def _aware_key(d: DocumentoEstructurado):
            f = d.fecha_acto
            if f is None:
                return datetime.max.replace(tzinfo=_tz.utc)
            if f.tzinfo is None:
                return f.replace(tzinfo=_tz.utc)
            return f

        return sorted(self.documentos, key=_aware_key)


class ArgumentoCandidato(BaseModel):
    regla_id: str
    descripcion: str
    cita_normativa_propuesta: str
    datos_disparo: dict[str, Any]
    impacto_estimado: Optional[str] = None


class ArgumentoVerificado(BaseModel):
    regla_id: str
    descripcion: str
    cita_verificada: str
    referencia_normativa_canonica: str
    confianza: float = Field(ge=0.0, le=1.0)
    datos_disparo: dict[str, Any]
    impacto_estimado: Optional[str] = None

File: frontend/src/types/defensia.ts
export type Tributo = "IRPF" | "IVA" | "ISD" | "ITP" | "PLUSVALIA";

export const TRIBUTO_LABELS: Record<Tributo, string> = {
  IRPF: "IRPF",
  IVA: "IVA",
  ISD: "Sucesiones y Donaciones",
  ITP: "Transmisiones Patrimoniales",
  PLUSVALIA: "Plusvalía Municipal",
};

export type Fase =
  | "COMPROBACION_REQUERIMIENTO"
  | "COMPROBACION_PROPUESTA"
  | "COMPROBACION_POST_ALEGACIONES"
  | "LIQUIDACION_FIRME_PLAZO_RECURSO"
  | "SANCIONADOR_INICIADO"
  | "SANCIONADOR_PROPUESTA"
  | "SANCIONADOR_IMPUESTA"
  | "REPOSICION_INTERPUESTA"
  | "TEAR_INTERPUESTA"
  | "TEAR_AMPLIACION_POSIBLE"
  | "FUERA_DE_ALCANCE"
  | "INDETERMINADA";

export const FASE_LABELS: Record<Fase, string> = {
  COMPROBACION_REQUERIMIENTO: "Requerimiento",
  COMPROBACION_PROPUESTA: "Propuesta de liquidación",
  COMPROBACION_POST_ALEGACIONES: "Alegaciones presentadas",
  LIQUIDACION_FIRME_PLAZO_RECURSO: "Plazo de recurso",
  SANCIONADOR_INICIADO: "Sancionador iniciado",
  SANCIONADOR_PROPUESTA: "Propuesta de sanción",
  SANCIONADOR_IMPUESTA: "Sanción impuesta",
  REPOSICION_INTERPUESTA: "Reposición interpuesta",
  TEAR_INTERPUESTA: "TEAR interpuesta",
  TEAR_AMPLIACION_POSIBLE: "Ampliación TEAR posible",
  FUERA_DE_ALCANCE: "Fuera de alcance",
  INDETERMINADA: "Indeterminada",
};

export type TipoDocumento =
  | "REQUERIMIENTO"
  | "PROPUESTA_LIQUIDACION"
  | "LIQUIDACION_PROVISIONAL"
  | "ACUERDO_INICIO_SANCIONADOR"
  | "PROPUESTA_SANCION"
  | "ACUERDO_IMPOSICION_SANCION"
  | "ESCRITO_ALEGACIONES_USUARIO"
  | "ESCRITO_REPOSICION_USUARIO"
  | "ESCRITO_RECLAMACION_TEAR_USUARIO"
  | "ACTA_INSPECCION"
  | "PROVIDENCIA_APREMIO"
  | "RESOLUCION_TEAR"
  | "RESOLUCION_TEAC"
  | "SENTENCIA_JUDICIAL"
  | "JUSTIFICANTE_PAGO"
  | "FACTURA"
  | "ESCRITURA"
  | "LIBRO_REGISTRO"
  | "OTROS";

export type EstadoExpediente =
  | "borrador"
  | "en_analisis"
  | "dictamen_listo"
  | "archivado";

export interface DocumentoEstructurado {
  id: string;
  nombre_original: string;
  tipo_documento: TipoDocumento;
  fecha_acto: string | null;
  datos: Record<string, unknown>;
  clasificacion_confianza: number;
}

export interface Expediente {
  id: string;
  titulo: string;
  tributo: Tributo;
  ccaa: string;
  estado: EstadoExpediente;
  fase_detectada: Fase;
  fase_confianza: number;
  num_documentos: number;
  dias_restantes: number | null;
  creado_en: string;
  actualizado_en: string;
}

export interface ExpedienteDetalle extends Expediente {
  documentos: DocumentoEstructurado[];
  brief: Brief | null;
  dictamen: Dictamen | null;
  escritos: Escrito[];
}

export interface Brief {
  id: string;
  texto: string;
  chat_history: Array<{ role: "user" | "assistant"; content: string }>;
  creado_en: string;
}

export interface ArgumentoVerificado {
  regla_id: string;
  descripcion: string;
  cita_verificada: string;
  referencia_normativa_canonica: string;
  confianza: number;
  datos_disparo: Record<string, unknown>;
  impacto_estimado: string | null;
}

export interface Dictamen {
  id: string;
  expediente_id: string;
  argumentos: ArgumentoVerificado[];
  resumen: string;
  creado_en: string;
}

export interface Escrito {
  id: string;
  expediente_id: string;
  plantilla: string;
  contenido_markdown: string;
  exportado: boolean;
  creado_en: string;
  actualizado_en: string;
}

export interface CuotaMensual {
  plan: "particular" | "autonomo" | "creator";
  usados: number;
  limite: number;
  restantes: number;
  extras_disponibles: number;
  precio_extra_eur: number;
  reset_en: string;
}
